### PR TITLE
core: use the same scoring curve for desktop in all channels

### LINKED ISSUE #9436: Make desktop perf metric score curves identical everywhere
When running tests on Lighthouse and PageSpeed Insight, we spotted some inconsistencies in the `speed index` classification of what is a 'fast', i.e. green/orange/red: 

- In Lighthouse, a `speed index` of 3.2s is green

lighthouse

- In Page Speed Insights, a `speed index` of 1.6s is orange

pagespeed_insights

Both results are observed in Desktop mode.

Since `speed index` is a Lab Data field, I expected the 2 tools to give the same classification.

### PR DESCRIPTION
**Pull request by connorjclark**
Fixes #9436.

The first inclination was to add some overrides in `config.js`, [kinda like this](https://github.com/GoogleChrome/lighthouse/blob/7637f0f69d978e3dcc8e88b744414d9c69065f71/lighthouse-core/config/config.js#L534-L549). But having the scoring curve numbers co-located with the audits is just so good, so let's try to keep them together.

The next inclination was to modify the audit `context` to have `hostDevice`. However, this was kind of awkward. For example, there's no great place to keep the logic of host agent checking, which is needed in creating the base artifacts and now here.

So instead of avoiding making a new artifact, I suggest we embrace it. This will be useful for #9713 too.

~~Draft to get consensus. I only did one audit.~~

### CODE REVIEW

File: lighthouse-core/audits/metrics/first-contentful-paint.js [hunk 1]
```diff
@@ -27,21 +27,31 @@ class FirstContentfulPaint extends Audit {
...
-  static get defaultOptions() {
-    return {
-      // 75th and 95th percentiles HTTPArchive -> median and PODR
-      // https://bigquery.cloud.google.com/table/httparchive:lighthouse.2018_04_01_mobile?pli=1
-      // see https://www.desmos.com/calculator/2t1ugwykrl
-      scorePODR: 2000,
-      scoreMedian: 4000,
-    };
+  static getDefaultScoreOptions(artifacts, context) {
+    if (context.settings.emulatedFormFactor === 'mobile' || artifacts.HostDevice === 'mobile') {
```
Comment (connorjclark) [inline lighthouse-core/audits/metrics/first-contentful-paint.js]: I think this logic is right. @paulirish can you confirm?

File: types/artifacts.d.ts [hunk 2]
```diff
@@ -29,6 +29,8 @@ declare global {
       LighthouseRunWarnings: string[];
       /** Whether the page was loaded on either a real or emulated mobile device. */
       TestedAsMobileDevice: boolean;
+      /** Device which Chrome is running on. */
+      HostDevice: 'desktop'|'mobile';
```
Comment (patrickhulce) [inline types/artifacts.d.ts]: 🚲 🏠 `HostFormFactor`?

File: lighthouse-core/audits/metrics/first-contentful-paint.js [hunk 3]
```diff
@@ -54,12 +64,13 @@ class FirstContentfulPaint extends Audit {
     const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
     const metricComputationData = {trace, devtoolsLog, settings: context.settings};
     const metricResult = await ComputedFcp.request(metricComputationData, context);
+    const scoreOptions = context.options || this.getDefaultScoreOptions(artifacts, context);
```
Comment (patrickhulce) [inline lighthouse-core/audits/metrics/first-contentful-paint.js]: can we have this be performed by the runner? i.e. check if default options is a function that accepts arguments and invoke it for the default options and/or permanently convert it to a function instead of getter property?

I like this idea of passing in context to determine default options overall though :)

File: types/artifacts.d.ts (hunk 2 above)
Comment (connorjclark) [reply types/artifacts.d.ts]: +1. What do others think?

File: lighthouse-core/audits/metrics/first-contentful-paint.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: > permanently convert it to a function instead of getter property?

That was another approach I started with. @paulirish and I thought it best to avoid changing the interface for default options, but I didn't consider supporting both a function and the (current) getter property, which I like.

Comment (patrickhulce) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: > thought it best to avoid changing the interface for default options

If we had a bunch of users of our default options already I'd agree, but we're still in the early stages there so if we'd like to move IMO now would be the time rather than indefinitely support two modes :)

Comment (brendankenny) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: > That was another approach I started with. @paulirish and I thought it best to avoid changing the interface for default options, but I didn't consider supporting both a function and the (current) getter property, which I like.

> If we had a bunch of users of our default options already I'd agree, but we're still in the early stages there so if we'd like to move IMO now would be the time rather than indefinitely support two modes :)

having a
```js
static getDefaultOptions(context) {
  if (artifacts.TestedAsMobileDevice) {
    return {
      scorePODR: 2000,
      scoreMedian: 4000,
    }
  } else {
    return {
      scorePODR: 800,
      scoreMedian: 1600,
    }
  }
}
}

static async audit(artifacts, context) {
  const scoreOptions = context.options;
  // ...
}
```
seems ergonomically (and functionally) equivalent to something like
```js
static get defaultOptions() {
  return {
    mobile: {
      scorePODR: 2000,
      scoreMedian: 4000,
    },
    desktop: {
      scorePODR: 800,
      scoreMedian: 1600,
    },
  };
}

static async audit(artifacts, context) {
  const scoreOptions = context.options[TestedAsMobileDevice ? 'mobile' : 'desktop'];
  // ...
}
```
and avoids churn for anyone using audit options.

Comment (brendankenny) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: but speaking of users of default options...is there any reason to keep score control points in the options? The main push for that was to be able to make a desktop config that defined its own score curves (#4873) to override the default ones, but if desktop scoring is being moved into the audits themselves, any reason to continue defining them circuitously like this instead of as regular properties?

File: lighthouse-core/audits/metrics/first-contentful-paint.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: Emulating desktop on a mobile device doesn't make much sense to me. Shouldn't those runs be scored with numbers tuned for a mobile device?

File: lighthouse-core/audits/metrics/first-contentful-paint.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: > any reason to continue defining them circuitously like this instead of as regular properties?

Keeping it in the options is easy and offers users a way to override our opinions on scoring. Seems valuable to me.

File: lighthouse-core/audits/metrics/first-contentful-paint.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: > Emulating desktop on a mobile device doesn't make much sense to me. Shouldn't those runs be scored with numbers tuned for a mobile device?

it doesn't make much sense why someone would be doing that to me either :) but if lantern doesn't correctly emulate desktop there it should be fixed at that level, not patched up at the scoring level

File: lighthouse-core/audits/metrics/first-contentful-paint.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: > offers users a way to override our opinions on scoring

my understanding is that's increasingly less desirable from the official metric audits, and if users want to do that they should use the computed metrics in custom audits.

Comment (connorjclark) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: tho I am assuming users actually use this

File: lighthouse-core/audits/metrics/first-contentful-paint.js (hunk 1 above)
Comment (paulirish) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: Yah I agree that `TestedAsMobileDevice` gives us the signal we want.  
real mobile hardware + desktop emulation is stupid but that's solved with a warning telling the developer they're stupid rather than mis-scoring them.   

Thank god we don't have first-class support for tablet with a mouse or whatever. :)

----

I still like having HostFormFactor in the artifacts tho.. even if we aren't using it here.
I guess that could be a separate PR, but no big deal.

File: lighthouse-core/audits/metrics/first-contentful-paint.js (hunk 3 above)
Comment (paulirish) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: > tho I am assuming _some_ user actually uses this

Looks like the only people using it are [folks that are reusing the desktop config](https://github.com/search?l=JavaScript&q=scorePODR+options+emulatedFormFactor&type=Code).



> but speaking of users of default options...is there any reason to keep score control points in the options? 

I was thinking the same thing.

I'm in favor of moving scoring out out of options. (Roughly reverting #4927)  If a user really wants different scoring, they can create a plugin that requires a core metric audit and then recomputes the score their way.

I'm happy to add friction there.  I want audit options to be used for all sorts of other things but core perf metric scoring doesn't seem like something that should be part of our extensibility story.

@patrickhulce how sad would you be?

File: types/artifacts.d.ts (hunk 2 above)
Comment (paulirish) [reply types/artifacts.d.ts]: +1 to hostformfactor

File: lighthouse-core/audits/metrics/first-contentful-paint.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: > Looks like the only people using it are folks that are reusing the desktop config.

A GitHub search would find any open source tools that use Lighthouse + these config options, but it doesn't tell us anything about how power users might be using configs in their infra to run Lighthouse.

Comment (patrickhulce) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: Setting it up with options allows folks to run with advanced throttling on different connection types and run with just a slightly modified config ala `lr-desktop-config`. To say that removing scoring from options adds friction feels like quite an understatement. If we removed, they would need to completely reimplement *every metric audit* just to have meaningful scores of any connection type outside our two blessed ones, which sounds like an effective way to completely kill testing of other connection types. 

We also originally structured it this way to offer metric values on different connection types. We had dreams once upon a time of offering multiple views within the same report and while no one has really talked about something like that recently, we get that for free when the inputs all come from options and it would be rather difficult to accomplish and/or require hard-coded copies without it. If #8374 had been able to land, I would have shipped my slow-3g performance plugin that does this :)

Nuking scoring in options kills both of these dreams, and I'm not personally convinced that a custom config isn't already enough friction to discourage it. AFAICT from my brief search there are exactly 0 usages of custom `scorePODR` in open source github where the values aren't just copied from our guidance or `lr-desktop-config`.

In summary, I would be very, very sad to see this die.

Comment (paulirish) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: okay. :)

we'll keep the scoring parameters in audit options.

Comment (brendankenny) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: I don't think it's a huge burden to make users do `class Slow3gInteractive extends Interactive {static getScoreOptions {return myAlteredScoreOptions;}`. But we could also be rid of the (undocumented) passing of an audit's default options back into itself with a simple `const {options = this.defaultOptions} = context` in the audit and still have the override behavior without `runner` trickiness.

...but I can live with the status quo :)

But I would very much like to see no new `options` magic for something we all agree no one uses to do something complicated when we can do it in a simple way :)

File: lighthouse-core/audits/metrics/first-contentful-paint.js [hunk 4]
```diff
@@ -54,12 +61,13 @@ class FirstContentfulPaint extends Audit {
...
     const metricComputationData = {trace, devtoolsLog, settings: context.settings};
     const metricResult = await ComputedFcp.request(metricComputationData, context);
+    const scoreOptions = context.options[artifacts.TestedAsMobileDevice ? 'mobile' : 'desktop'];
 
     return {
       score: Audit.computeLogNormalScore(
         metricResult.timing,
-        context.options.scorePODR,
-        context.options.scoreMedian
+        scoreOptions.scorePODR,
```
Comment (connorjclark) [inline lighthouse-core/audits/metrics/first-contentful-paint.js]: @patrickhulce @paulirish @brendankenny Good?

Comment (patrickhulce) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: probably should still note this as a breaking change in release notes, but this WFM 👍

Comment (connorjclark) [reply lighthouse-core/audits/metrics/first-contentful-paint.js]: agreed. we should adopt a way to signify breaking changes for release-notes-writing time.

File: lighthouse-core/test/audits/metrics/first-meaningful-paint-test.js [hunk 5]
```diff
@@ -22,7 +22,7 @@ describe('Performance: first-meaningful-paint audit', () => {
     const context = {options, settings: {throttlingMethod: 'provided'}, computedCache: new Map()};
     const fmpResult = await FMPAudit.audit(artifacts, context);
 
-    assert.equal(fmpResult.score, 1);
+    assert.equal(fmpResult.score, 0.96);
```
Comment (paulirish) [inline lighthouse-core/test/audits/metrics/first-meaningful-paint-test.js]: kinda weird to see all these scores change in the tests.

im assuming thats because `TestedAsMobileDevice` is falsy by default in these tests.
that's a little odd as LH is is otherwise mobile by default.

we need something better here. options i can think of:

- add `{TestedAsMobileDevice: true}` into all these tests to make it explicit
- change the ternary in the metric audits to check if `TestedAsMobileDevice === false ? 'desktop' : 'mobile'`  because i guess it'd be `undefined` in all these test scenarios?

Comment (connorjclark) [reply lighthouse-core/test/audits/metrics/first-meaningful-paint-test.js]: I considered both of those.

the first was annoying and a big change to the tests

the second is ehhhh. I would do this if only there was one single place this condition lived. but it's in many places

but
> that's a little odd as LH is is otherwise mobile by default.
is so convincing that we should do one of these. I choose the second. is there a good place to move this code to a function?

Comment (connorjclark) [reply lighthouse-core/test/audits/metrics/first-meaningful-paint-test.js]: ehhhhh i just did 2)
